### PR TITLE
Force delete flag

### PR DESCRIPTION
**Pull request by KirilKabakchiev**
Adds a ?force=true/false flag to SM APIs that allows ignoring around tx errors. This allows deleting instances/bindings if the broker is stopped/missing/returns errors

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func (s *Settings) Validate() error {
 }
 
 type Options struct {
-	Repository        storage.TransactionalRepository
+	Repository        *storage.InterceptableTransactionalRepository
```
Comment (NickyMateev) [inline api/api.go]: This can stay `storage.TransactionalRepository`

File: api/base_controller.go [hunk 2]
```diff
@@ -50,7 +50,7 @@ type BaseController struct {
 
 	resourceBaseURL string
 	objectType      types.ObjectType
-	repository      storage.Repository
+	repository      *storage.InterceptableTransactionalRepository
```
Comment (NickyMateev) [inline api/base_controller.go]: `storage.Repository`

File: storage/interceptable_repository.go [hunk 3]
```diff
@@ -232,6 +232,35 @@ func (ir *queryScopedInterceptableRepository) Delete(ctx context.Context, object
 	return nil
 }
 
+func (ir *queryScopedInterceptableRepository) ForceDelete(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) error {
```
Comment (NickyMateev) [inline storage/interceptable_repository.go]: Spare the code duplication by extracting the common logic and refactoring the `Delete`, `DeleteReturning` and `ForceDelete` methods like so:

```go
func (ir *queryScopedInterceptableRepository) DeleteReturning(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) (types.ObjectList, error) {
	var resultList types.ObjectList
	deleteObjectFunc := func(ctx context.Context, _ Repository, _ types.ObjectList, deletionCriteria ...query.Criterion) error {
		var err error
		if resultList, err = ir.repositoryInTransaction.DeleteReturning(ctx, objectType, deletionCriteria...); err != nil {
			return err
		}
		return nil
	}

	if err := ir.delete(ctx, objectType, criteria, deleteObjectFunc); err != nil {
		return nil, err
	}

	return resultList, nil
}

func (ir *queryScopedInterceptableRepository) Delete(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) error {
	deleteObjectFunc := func(ctx context.Context, _ Repository, _ types.ObjectList, deletionCriteria ...query.Criterion) error {
		if err := ir.repositoryInTransaction.Delete(ctx, objectType, deletionCriteria...); err != nil {
			return err
		}
		return nil
	}

	return ir.delete(ctx, objectType, criteria, deleteObjectFunc)
}

func (ir *queryScopedInterceptableRepository) ForceDelete(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) error {
	deleteObjectFunc := func(ctx context.Context, _ Repository, _ types.ObjectList, deletionCriteria ...query.Criterion) error {
		if err := ir.repositoryInTransaction.ForceDelete(ctx, objectType, deletionCriteria...); err != nil {
			return err
		}
		return nil
	}

	return ir.delete(ctx, objectType, criteria, deleteObjectFunc)
}

func (ir *queryScopedInterceptableRepository) delete(ctx context.Context, objectType types.ObjectType, criteria []query.Criterion, deleteObjectFunc InterceptDeleteOnTxFunc) error {
	if deleteOnTxFunc, found := ir.deleteOnTxFuncs[objectType]; found {
		objects, err := ir.List(ctx, objectType, criteria...)
		if err != nil {
			return err
		}
		delete(ir.deleteOnTxFuncs, objectType)
		if err := deleteOnTxFunc(deleteObjectFunc)(ctx, ir, objects, criteria...); err != nil {
			ir.deleteOnTxFuncs[objectType] = deleteOnTxFunc
			return err
		}
		ir.deleteOnTxFuncs[objectType] = deleteOnTxFunc

	} else {
		if err := deleteObjectFunc(ctx, nil, nil, criteria...); err != nil {
			return err
		}
	}

	return nil
}
```

File: test/service_binding_test/service_binding_test.go [hunk 4]
```diff
@@ -1663,6 +1671,25 @@ var _ = DescribeTestsFor(TestCase{
 										Ready: true,
 									})
 								})
+
+								When("unbind is forceful", func() {
+									It("deletes the binding and marks the operation with success", func() {
+										resp := forceDeleteBinding(ctx.SMWithOAuthForTenant, testCase.async, testCase.expectedDeleteSuccessStatusCode)
+
+										instanceID, _ = VerifyOperationExists(ctx, resp.Header("Location").Raw(), OperationExpectations{
```
Comment (NickyMateev) [inline test/service_binding_test/service_binding_test.go]: bindingID

File: test/service_binding_test/service_binding_test.go [hunk 5]
```diff
@@ -1663,6 +1671,25 @@ var _ = DescribeTestsFor(TestCase{
...
+											Category:          types.DELETE,
+											State:             types.SUCCEEDED,
+											ResourceType:      types.ServiceBindingType,
+											Reschedulable:     false,
+											DeletionScheduled: false,
+										})
+
+										VerifyResourceDoesNotExist(ctx.SMWithOAuthForTenant, ResourceExpectations{
+											ID:   instanceID,
+											Type: types.ServiceInstanceType,
```
Comment (NickyMateev) [inline test/service_binding_test/service_binding_test.go]: `types.ServiceBindingType`

**Comment by KirilKabakchiev** [comment]
implementation has been postponed

**Pull request closed without merging**